### ISSUE TITLE
[V3] The exampleValue parameter within fromProviderState does not accept non-string data types

### ISSUE THREAD

**Issue by lviana-menlosecurity** [pact-foundation/pact-js#696]
### Software versions

_Please provide at least OS and version of pact-js_

- **OS**: _e.g. Mac OSX 10.11.5_
- **Consumer Pact library**: "@pact-foundation/pact": "^10.0.0-beta.37"
- **Provider Pact library**: "@pact-foundation/pact": "^10.0.0-beta.37"
- **Node Version**: 14.16.1

### Issue Checklist

Please confirm the following:

- [X] I have upgraded to the latest
- [X] I have the read the FAQs in the Readme
- [X] I have triple checked, that there are **no unhandled promises** in my code
- [ ] I have set my log level to debug and attached a log file showing the complete request/response cycle
- [ ] For bonus points and virtual high fives, I have created a reproduceable git repository (see below) to illustrate the problem

### Expected behaviour

The exampleValue parameter within fromProviderState should accept more than only string data types for TypeScript tests, like the number in the JavaScript example below:

https://github.com/pact-foundation/pact-js/blob/feat/v3.0.0/examples/v3/provider-state-injected/consumer/transaction-service.test.js#L45

### Actual behaviour

Currently, exampleValue is defined as string:
https://github.com/pact-foundation/pact-js/blob/feat/v3.0.0/src/v3/matchers.ts#L450

This prevents us from using non-string data types within TypeScript tests, as that will fail. The following needs to happen:
1. The exampleValue parameter data type should not be limited to only strings.
2. The replacement code in the provider test side needs to honor the data type in the replacement. I say that because I avoided problem (1) above by changing the exampleValue data type from "string" to "any" and that allowed the consumer tests to pass, but the provider tests still failed because it still replaced my variable with a string instead of number (even though I return a number in stateHandlers) and our provider expects the field to be a number.

**Comment by mefellows** [comment]
Just noting that I've fixed (1) but not (2).

Release beta 39 contains the change: https://github.com/pact-foundation/pact-js/releases/tag/v10.0.0-beta.39

(2) is a bit more difficult, because we would need to infer the return types from the contract. I think it's doable, but we don't currently have any specification to document how we do this.

@uglyog any thoughts? The request in (2) is to wrap some type safety around the provider injected parameters.

**Comment by lviana-menlosecurity** [comment]
Thanks for fixing this, @mefellows. However, after the ProviderStateInjectedValue type was updated, that ended up breaking things elsewhere. For example, using fromProviderState on a request path is now broken because it expects a type of MatchersV3.Matcher<string> on the line below:
https://github.com/pact-foundation/pact-js/blob/feat/v3.0.0/src/v3/pact.ts#L51

**Comment by TimothyJones** [comment]
I'm just heading to bed here, but I think something like

```
export interface ProviderStateInjectedValue<T> extends Matcher<T> {
  expression: string;
}
export function fromProviderState<V extends AnyJson>(
  expression: string,
  exampleValue: V
) ProviderStateInjectedValue<V>
```

_might_ fix this.

**Comment by lviana-menlosecurity** [comment]
@TimothyJones , I tried your suggested change above locally, and it indeed seems to fix the problem with using fromProviderState on a request path . ;-)
We still need to fix issue (2) within the replacement code so that the provider test honors the data type in the replacement.

**Comment by uglyog** [comment]
In Pact-JVM I have an associated enum to record the data type conversion that is required. See https://github.com/pact-foundation/pact-jvm/blob/master/core/support/src/main/kotlin/au/com/dius/pact/core/support/expressions/ExpressionParser.kt#L9

These are set in the DSL, were the required type is known.

However, only two conversions ended up being actually needed: STRING for things like paths, query parameters and then RAW which means do not convert it. You could get away with just a boolean value to indicate if the value needs to be converted to a string.

**Comment by TimothyJones** [comment]
I'm not sure I fully understand the typesafety problem in number 2 - could you share a code example where you're experiencing the problem?

It might be hard to get typesafety correct, since typescript is compile time, and the types given to the state handlers aren't known until runtime.

**Comment by mefellows** [comment]
(forgot to hit send on this, but agree with Tim).

Great pickup Tim. I'll get another minor change out today with that @lviana-menlosecurity.

_UPDATE_: change is out in the `beta.40` release.

> We still need to fix issue (2) within the replacement code so that the provider test honors the data type in the replacement.

This can be left open as an enhancement. The test will fail in either instance - either the wrong type is sent back or the wrong data. Either way, test is failing and somebody needs to look. 

Because the types will be encoded in JSON I'm not sure how helpful that is actually going to be. We'd need to actually encode more information into the contract (which there are no plans to do) to do much better than logging a warning or something intelligent to help with debugging.

**Comment by lviana-menlosecurity** [comment]
@TimothyJones, here is a code example you asked for.
In the consumer test, we have something like the following:
```
      provider
        .given('a policy revision exists')
        .uponReceiving('a rules request to add')
        .withRequest({
          method: 'PUT',
          path: MatchersV3.fromProviderState('/api/policy/v1/rules/risks/${staged_rev_id}', '/api/policy/v1/rules/risks/788ed3db193b4e1aa33e380da5294853'),
          headers: { 'Content-Type': 'application/json' },
          body: {
            rev_id: MatchersV3.fromProviderState('${staged_rev_id}', '788ed3db193b4e1aa33e380da5294853'),
            updated_at: MatchersV3.fromProviderState('${updated_at}', 1623700550085)
          }
        })
        .willRespondWith({
          status: 200
        });
```
Notice how the body gets 2 values from the state: rev_id is a string (since the exampleValue is wrapped in quotes), and updated_at is a number (no quotes). With the commit above, which is part of the beta.40 release, the consumer test now passes, since we can now use AnyJson type for the exampleValue.
However, the provider test, which looks like the following, still fails:
```
      stateHandlers: {
        'a policy revision exists': () => {
          return {
            staged_rev_id: 'e3a8b509e60b46baa0ab9044a42820b2',
            updated_at: 1623765384870 // Even though a number is passed, Pact still uses a string in the actual request
          }
        },
      },
```
Despite the fact that a number is returned for updated_at in the stateHandlers above, Pact still uses a string in the actual request. However, our provider expects a number for that field, and the request fails if the type is incorrect. This is the problem.

**Comment by mefellows** [comment]
Thanks, that makes more sense and seems like a bug on the surface. I'll have to do a little more digging (this will be in the shared core, I believe).

**Comment by mefellows** [comment]
As part of refactoring #702, I've demonstrated the issue is in the rust core:

```
[2021-06-25T11:40:54Z DEBUG reqwest::async_impl::client] response '200 OK' for http://localhost:50110/_pactSetup
[2021-06-25T11:40:54Z DEBUG pact_verifier::provider_client] State change request: Response { url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(50110), path: "/_pactSetup", query: None, fragment: None }, status: 200, headers: {"x-powered-by": "Express", "content-type": "application/json; charset=utf-8", "content-length": "23", "etag": "W/\"17-nI2ZGx8NwIYvqPResGsg7KwEQ/k\"", "date": "Fri, 25 Jun 2021 11:40:54 GMT", "connection": "keep-alive"} }
[2021-06-25T11:40:54Z DEBUG pact_verifier] State Change: "ProviderState { name: "Account Test001 exists", params: {"accountRef": String("Test001")} }" -> Ok({"accountNumber": Number(96827)})
[2021-06-25T11:40:54Z INFO  pact_verifier] Running provider verification for 'a request to get the account details via POST'
[2021-06-25T11:40:54Z DEBUG pact_matching] Applying body generators...
[2021-06-25T11:40:54Z DEBUG pact_matching::models::generators] apply_body_generators: JSON content type
[2021-06-25T11:40:54Z DEBUG pact_matching::models::generators] Applying generator ProviderStateGenerator("${accountNumber}", None) to key $.accountNumber
[2021-06-25T11:40:54Z DEBUG pact_matching::json] Generating value from ProviderStateGenerator("${accountNumber}", None) with context {"accountNumber": Number(96827)}
[2021-06-25T11:40:54Z DEBUG pact_matching::json] Generated value = Ok(String("96827"))
[2021-06-25T11:40:54Z INFO  pact_verifier::provider_client] Sending request to provider at http://localhost:50110/
[2021-06-25T11:40:54Z DEBUG pact_verifier::provider_client] Provider details = ProviderInfo { name: "Account Service", protocol: "http", host: "localhost", port: Some(50110), path: "/" }
[2021-06-25T11:40:54Z DEBUG pact_verifier::provider_client] Sending request Request ( method: POST, path: /accounts/search/findOneByAccountNumberIdInBody, query: None, headers: Some({"Content-Type": ["application/json; charset=utf-8"], "Accept": ["application/hal+json"]}), body: Present(25 bytes, application/json) )
[2021-06-25T11:40:54Z TRACE pact_verifier::provider_client] body: {"accountNumber":"96827"}
```

(mostly for my future reference).

TL;DR - the correct shape goes in, but things seem to be stringified on the way back out. I'll leave this open to track it.
